allennlp/nn/beam_search.py

```diff
@@ -222,6 +223,11 @@ def _search(
                 start_predicted_classes,
             ) = start_class_log_probabilities.topk(self.beam_size)
 
+        if isinstance(self.sampler, GumbelMaxSampler):
+            last_true_probabilities = torch.gather(
+                start_class_log_probabilities, 1, start_predicted_classes
+            )
```

Comment:
I actually think we should always gather `last_true_log_probabilities` since there is no guarantee with other samplers that they won't perturb the log probabilities.

```suggestion
        # Some samplers add perturb the log probabilities with random noise, such
        # as the GumbelMaxSampler, so we need to recover and keep track of the
        # true log probabilities here.
        last_true_log_probabilities = torch.gather(
            start_class_log_probabilities, 1, start_predicted_classes
        )
```

Would need a similar change below as well.

Reply:
Should I still check `if self.sampler is not None:` ? Right now only the gumbel sampler will return perturbed log probabilities, and deterministic beam search will always keep track of the true probabilities

Reply:
Ah true, yes let's do that